src/bootstrap/src/core/builder.rs

```diff
@@ -1630,10 +1656,19 @@ impl<'a> Builder<'a> {
             )
             .env("RUSTC_ERROR_METADATA_DST", self.extended_error_dir())
             .env("RUSTC_BREAK_ON_ICE", "1");
-        // Clippy support is a hack and uses the default `cargo-clippy` in path.
-        // Don't override RUSTC so that the `cargo-clippy` in path will be run.
-        if cmd != "clippy" {
-            cargo.env("RUSTC", self.bootstrap_out.join("rustc"));
+
+        // Set RUSTC_WRAPPER to the bootstrap shim, which switches between beta and in-tree
+        // sysroot depending on whether we're building build scripts.
+        // NOTE: we intentionally use RUSTC_WRAPPER so that we can support clippy - RUSTC is not
+        // respected by clippy-driver; RUSTC_WRAPPER happens earlier, before clippy runs.
+        cargo.env("RUSTC_WRAPPER", self.bootstrap_out.join("rustc"));
+        // NOTE: we also need to set RUSTC so cargo can run `rustc -vV`; apparently that ignores RUSTC_WRAPPER >:(
+        cargo.env("RUSTC", self.bootstrap_out.join("rustc"));
+
+        // Someone might have set some previous rustc wrapper (e.g.
+        // sccache) before bootstrap overrode it. Respect that variable.
+        if let Some(existing_wrapper) = env::var_os("RUSTC_WRAPPER") {
+            cargo.env("RUSTC_WRAPPER_REAL", existing_wrapper);
```

Comment:
cc @the8472

Reply:
I haven't touched this code, have I? 😅
But yeah, I have that in my ENV by default. Should I test it now or just yell later if something breaks?

Reply:
You might also want to give it a more descriptive name? Perhaps BOOTSTRAP_SHIM_RUSTC_WRAPPER.

Reply:
i tagged you because of https://github.com/rust-lang/rust/pull/110854#issuecomment-1523819609, wanted to let you know this shouldn't break your config :)

all these variables use `_REAL`, they aren't public facing. open an issue or pr if you think it should change.